### PR TITLE
Switch over to use the higher level JSON APIs that handle BOM automatically

### PR DESCRIPTION
**Pull request by danielmarbach**
Follow up for https://github.com/Particular/NServiceBus/pull/6768

### CODE REVIEW

**Comment by danielmarbach** [comment]
@SimonCropp do you remember why your version went with the lowest level approach? Using the UTF8Writer and Reader is usually the approach when you use custom parsers that are interested in the tokens. For the serializer it seems this approach has some downsides as for example BOM handling is not automatic.

From my point of view, I cannot recall any special case that could not be handled over the JsonSerializerOptions that today would be enabled by exposing the reader and writer options. On the contrary, exposing those options exposes to a potential misconfiguration between the reader/writer options and the json serializer options. 

What do you think?

File: src/NServiceBus.Core/Serializers/XML/ReadOnlyMemoryStream.cs [hunk 1]
```diff
@@ -1,16 +1,14 @@
-﻿// unset
```
Comment (DavidBoike) [inline src/NServiceBus.Core/Serializers/XML/ReadOnlyMemoryStream.cs:1]: If this is going to be used by JSON as well, should it be moved perhaps up a level to the NSB.Core/Serializers directory?

File: src/NServiceBus.Core/Serializers/XML/ReadOnlyMemoryStream.cs [hunk 2]
```diff
@@ -1,16 +1,14 @@
...
     using System;
     using System.IO;
 
-    class ReadOnlyStream : Stream
+    sealed class ReadOnlyMemoryStream : MemoryStream
     {
-        ReadOnlyMemory<byte> memory;
-        long position;
+        readonly ReadOnlyMemory<byte> memory;
+        int position;
```
Comment (DavidBoike) [inline src/NServiceBus.Core/Serializers/XML/ReadOnlyMemoryStream.cs:9]: Why this change? ReadOnlyMemory can't be longer than int.MaxValue bytes?

**Review by DavidBoike — APPROVED** [review]

File: src/NServiceBus.Core/Serializers/XML/ReadOnlyMemoryStream.cs (hunk 2 above)
Comment (danielmarbach) [reply src/NServiceBus.Core/Serializers/XML/ReadOnlyMemoryStream.cs:9]: Previously the implementation did cast the position to int on every Read call. This changed it to treat it int all the time on the field level but the outcome should be the same right?

Will double check though the runtime implementation again (I think I did it before but it doesn't hurt to be extra careful)

Comment (danielmarbach) [reply src/NServiceBus.Core/Serializers/XML/ReadOnlyMemoryStream.cs:9]: @DavidBoike example https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/IO/MemoryStream.cs#L25-L27

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/IO/MemoryStream.cs#L315

It does a bit more validation but :shrug:

**Comment by SimonCropp** [comment]
@danielmarbach it was "just a case" since i wasnt sure what people would want to configure, given it was a new serializer at the time.

for NSB i think it makes sense to limit the api surface area, and grow it in the future if anyone comes up with a legit use case in the future

**Review by andreasohlund — APPROVED** [review]

**Comment by andreasohlund** [comment]
Thanks for the input @SimonCropp I'll add a section [to the upgrade guide](https://github.com/Particular/docs.particular.net/pull/6089/files#diff-0d486f0afab953c430770c3f8114b3c1d561429c97e705cec67ae60b132a9443R11) that covers this change

**Comment by ramonsmits** [comment]
@danielmarbach This approach copies the data and then passed that to the deserializer method while the previous version didn't do it. Isn't this allocating more memory?

**Comment by danielmarbach** [comment]
The memory is [amortized because it is rented from the shared array pool](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadBufferState.cs#L25). There is copying overhead that is true, but I think that is an acceptable price in comparison of:

- Having to deal in a custom way with something the framework already does
- Having to expose on the settings the underlying low-level mechanics of configuring the reader and writer.
